### PR TITLE
Add field type blob to entity generator, fix #1651

### LINKED ISSUE #1651: Add new type Blob to entity generator
Working on a PR

### PR DESCRIPTION
**Pull request by andidev**
- works with databaseType = sql and mongodb
- supports required, minbytes, maxbytes validation (both front-end angular validation and back-end java bean validation)
- uses REST CRUD as any other field type generated (data is passed in the json obejct representing the entity)
- uses [ng-file-upload](https://github.com/danialfarid/ng-file-upload) angular directive to upload blobs as files (drag and drop support)

Tested file sizes over 200MB, there seems to be some limit around 250MB (javascriot FileReader seems to fail). Maybe there should be maxbytes limit set by default. Is it possible to release this with a BETA tag anyway?

### CODE REVIEW

File: app/templates/src/main/webapp/scripts/components/form/_maxbytes.directive.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/* globals $ */
+'use strict';
+
+angular.module('jhipsterApp')
```
Comment (jdubois) [inline app/templates/src/main/webapp/scripts/components/form/_maxbytes.directive.js]: it should be the user's app name : <%=angularAppName%>

File: app/templates/src/main/webapp/scripts/components/form/_minbytes.directive.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/* globals $ */
+'use strict';
+
+angular.module('jhipsterApp')
```
Comment (jdubois) [inline app/templates/src/main/webapp/scripts/components/form/_minbytes.directive.js]: Same thing here: <%=angularAppName%>

**Comment by jdubois** [comment]
@andidev I'm currently merging this, it's really great!!
There will be some work to do for big blobs: for varchars, for example, I set up the max column length in the liquibase changelog. Here I don't think this is possible, but we should document how you can handle bigger files (I currently have the issue with an image).
Anyway, that's really excellent!!!!

**Comment by andidev** [comment]
@jdubois 
How large is the image? I only encountered problems with files over 250MB.

I'm not sure this is the best way to handle blobs actually. Probably it is a better idea to pass some id in the json instead of the data and then let the data exist in a separate database table or on the disk and upload it before the entity is created with springs MultipartFile. Though that solution is a lot more complex and with this solution it is still possible to do that by using the entity generator and create a separate entity to hold the data.

Weather or not this is the best solution I think that it is a good start. I'm open for suggestions on all sort of changes. I'm gonna use this my self in the project I'm working on so I'm interested in making it better.

**Comment by jdubois** [comment]
Something like 100ko. But this will depend on the database, its configuration... So maybe we should document it.
Anyway, it's better to have it done quickly, with a simple solution, and we'll see later how we can improve it.

**Comment by deepu105** [comment]
Im doing it with ng flow and saving the document in the filesystem, im planning to have just the filepath stored in the entity

Im using the sample jhipster app provided by @danielshiplett
https://github.com/danielshiplett/jhipster_uploader

**Comment by danielshiplett** [comment]
In case you are running your JHipster application behind a proxy (haproxy, nginx, apache), beware that they may also contribute to the file size limit on upload.  Assuming the upload is in one chunk.

I know NGINX has a default upload size that is small.  250MB reminds me of a default value I've seen before.  Though I can quite remember which application.

**Comment by deepu105** [comment]
Angular upload looks simpler than ng flow and supports csrf and stuff so im gonna switch to angular upload i guess coz facing issues with ng flow

**Comment by andidev** [comment]
The limit around 250mb I encountered encountered was in the javascript FileReader. But probably there could be other limits as well. 

@deepu105 
How will you solve the issue where the user regrets creating an entity after uploading the image?

**Comment by deepu105** [comment]
@andidev i have just switched to angular upload which you used. Today ill try out this scenario. My current tgought process is to get the file as multipart in a spring resource along witg remaining parameters of the entity and use a service layer action to store the multipart to file system and path and other details as entity to db. Do im not gonna do an upload as soon as user selects a file, ill upload only when user hits save.

**Comment by andidev** [comment]
Sounds good! Did not think of that option.

**Comment by deepu105** [comment]
@andidev I got it working successfully, Im now posting the blob along with other entity filed as a multpart request and on the server side i save the file to file system first and then save the entity with the file path reference. since its asynchronous Im showing upload progressbar on the UI if user is still on it. Let me know if you are interested in this approach ill post my code snippets.
btw I have a query why did you write min bytes and maxbytes directives instead of using `ngf-min-size` and `ngf-max-size` ?

**Comment by deepu105** [comment]
btw angular upload is far better and simpler than ng-flow for simple uploads :+1:  flow might be better for chunk uploads and stuff like that

**Comment by andidev** [comment]
@deepu105 
Great, please post the snippets and I'll have a look at it. @jdubois what do you say?

I did not use `ngf-min-size` and `ngf-max-size` the validation does not use angular validation and it would require a adding code in the controller for making it work as the current validation works (if it's even possible).

One of the problem I encountered was when selecting a file smaller then the ngf-min-size the file would not be loaded and the validation message would be displayed. The problem here is that then you have a validation message showing that file is to small when the file is not even selected. This is okay but how can this validation state be distinguished from the required state validation (whereas the require validation uses angular validate so both will be shown at the same time I think)?

There where other problems as well and I found out the most consistent solution was to create a custom angular validate directive for min and max bytes. I think that maybe `nfg-min/max-size` makes more sense to use when uploading the file content at file select, but not otherwise. Please correct me if I'm wrong, maybe I missed something. Did you use `ngf-min/max-size` successfully?

**Comment by deepu105** [comment]
ill post them soon, need to remove some client specific code.but im not sure if this is the best way to do it, guess you guys can improve it

**Comment by deepu105** [comment]
@andidev Since my requirements were specific i have done many non generic stuff here which could be avoided/improved. But anyway take a look and let me know what you think, Im not very good in spring so the server side could be greatly improved

# controller
```javascript
.controller('DocumentViewController', function($scope, Principal, $state, $modalInstance, obj, Document, Company, UploadService, AlertService) {

        //accepted file types, should not be any space in the text, seperate types by comma
        $scope.accept = 'image/*,audio/*,video/*,application/pdf,application/msword,application/rtf,application/zip,application/x-zip-compressed,application/x-zip,text/csv,application/vnd.ms-excel,application/vnd.ms-powerpoint';
        $scope.rejFiles = null;
       

        $scope.file = {
            loading: false,
            start: false
        }

        $scope.cancel = function() {
            $modalInstance.dismiss('cancel');
        };

        $scope.editing = false;
        if (obj) {
            $scope.new = false;
            $scope.view = true;
            $scope.document = obj;
        } else {
            $scope.new = true;
            $scope.view = false;
            $scope.editing = true;
            $scope.files = null;
            $scope.document = {};
        }


        $scope.edit = function() {
            $scope.editing = true;
            $scope.editForm.fileSelected.$setValidity("required", true);
        };

        $scope.clear = function () {
            $scope.document = {docName: null, description: null, classification: null, filePath: null, fileSize: null, fileExtn: null, id: null};
            $scope.files = null;
            $scope.file.name = null;
            $scope.editForm.$setPristine();
            $scope.editForm.$setUntouched();
        };


        $scope.reset = function () {
            $scope.file.upload = null;
            $scope.file.loading = false,
            $scope.file.start = false;
            $scope.editForm.$submitted = false;
        }

        $scope.setFileName = function () {
            if ($scope.files && $scope.files.length) {
                $scope.file.name = $scope.files[0].name;
            }
        }

        $scope.$watch('files', function(newValue, oldValue) {
            if(newValue && newValue!== null && newValue !== ""){
                $scope.editForm.fileSelected.$setValidity("required", true);
            }
        });

        $scope.$watch('rejFiles', function(newValue, oldValue) {
            if(newValue !== oldValue && newValue.length > 0){
                AlertService.warning("Invalid File type selected. Allowed types are Image, Video, Audio, documents or Zip files");
            }
        });

        $scope.save = function () {
            if ($scope.document && $scope.document.id != null) {

                Document.update($scope.document,
                    function () {
                        $scope.clear();
                        $scope.reset();
                        $modalInstance.close(true);
                    });
            } else {
                if ($scope.files && $scope.files.length) {
                    $scope.upload($scope.files[0]);
                }
            }
        }

        $scope.upload = function (file) {

            $scope.file.upload = UploadService.upload(file, $scope.document);

            $scope.file.upload.progress(function (evt) {
                $scope.file = UploadService.mapEvent(evt);
                //TODO sent the progress hook to notification service
            }).success(function (data, status, headers, config) {
                $scope.file.loading = false;
                $scope.file.status = status;
                $scope.clear();
                $scope.reset();
                $modalInstance.close(true);
                //TODO complete the progress hook in notification service
            }).error(function (data, status, headers, config) {
                console.log('error status: ' + status);
                //$scope.reset();
                //TODO handle error
            });

        };

        $scope.cancelUpload = function(){
            if ($scope.file && $scope.file.upload) {
                $scope.file.upload.abort();
            }

            $scope.clear();
            $scope.reset();
            $scope.cancel();
        }
    });
```
# HTML

```html
              <div class="row" ng-show="new" >

                    <div class="col-md-12 form-group">
                        <label>File to Upload</label>
                        <input type="text" class="form-control" name="fileSelected" id="fileSelected"
                               ng-model="fileSelected" required style="display: none">

                        <div class="row pad">
                            <div class="col-sm-2 pad">
                                <button class="btn btn-default" ng-model-rejected="rejFiles" accept="{{accept}}" ngf-accept="accept" ngf-multiple="false" ngf-allow-dir="false" ngf-select ng-model="files" ngf-change="setFileName()">Select File</button>
                                <b>OR</b>
                            </div>
                            <div ngf-drop ngf-select ng-model="files" ng-model-rejected="rejFiles" accept="{{accept}}" ngf-accept="accept" class="drop-box col-sm-10" ngf-change="setFileName()"
                                 ngf-drag-over-class="drop-box-drag" ngf-multiple="false" ngf-allow-dir="false">
                                <span>Drag And Drop file here<br>
                                    <p ng-if="files.length > 0">File selected: <kbd>{{file.name}}</kbd></p>
                                </span>
                            </div>
                            <div ngf-no-file-drop>File Drag/Drop is not supported for this browser</div>
                        </div>
                        <div ng-show="editForm.fileSelected.$invalid">
                            <p class="help-block"
                               ng-show="editForm.fileSelected.$error.required">
                                This field is required.
                            </p>
                        </div>
                        <br/>

                        <br/>
                        <div ng-if="file.start" class="pad">
                            <div class="gallery-box">
                                <progressbar min="0" max="100" value="file.progress" class="progress-striped" ng-class="{active: file.loading}" type="info">
                                    <span>{{file.progress}}% Complete</span>
                                </progressbar>
                                <div class="btn-group" ng-if="file.loading">
                                    <a class="btn btn-xs btn-danger" ng-click="cancelUpload()">
                                        Cancel
                                    </a>
                                </div>
                            </div>
                            <div class="clear"></div>
                        </div>

                    </div>
                </div>
```
# css for drop
```css
.drop-box {
    border: 2px dashed #B4ABAB;
    height: 180px;
    text-align: center;
    vertical-align: middle;
    line-height: 180px;
}
.drop-box span {
    display: inline-block;
    vertical-align: middle;
    line-height: normal;
}
.drop-box-drag{
    border: 2px dashed #08DB6C !important;
}
kbd:empty {
    display: none;
}
```
# service
```javascript
'use strict';

angular.module('*****')
    .factory('Document', function ($resource, DateUtils) {
        return $resource('api/documents/:id', {}, {
            'query': { method: 'GET', isArray: true},
            'get': {
                method: 'GET',
                transformResponse: function (data) {
                    data = angular.fromJson(data);
                    return data;
                }
            },
            'update': { method:'PUT' }
        });
    })
    .factory('UploadService', function (Upload) {
        return {
            mapEvent: function (evt) {
                return {
                    total: evt.total,
                    loaded: evt.loaded,
                    progress : parseInt(100.0 * evt.loaded / evt.total),
                    name : evt.config.file.name,
                    loading : evt.total == evt.loaded? false : true,
                    start : true
                };
            },
            upload: function(file, doc){
                return Upload.upload({
                    url: 'api/documents/upload',
                    fields: doc,
                    file: file
                });
            }
        };
    });
```
# resource 
(it should be possible to upload the entire document object along with the file as two @RequestPart parameters, but for some reason it wasnt working for me so i passed as params, since i didnt have time to debug the issue )
```java
private static final int BUFFER_SIZE = 4096;

    /**
     * POST  /documents -> upload and create a new document.
     */
    @RequestMapping(value = "/documents/upload",
            method = RequestMethod.POST,
            produces = MediaType.APPLICATION_JSON_VALUE)
    @Timed
    @Secured(SECTION)
    public ResponseEntity<Void> upload(@RequestParam( "file") MultipartFile file, @RequestParam(value = "showInDashboard", required = false) boolean showInDashboard,
                                       @RequestParam("docName") String docName,@RequestParam(value = "description", required = false) String description,
                                       @RequestParam("classification") String classification, @RequestParam(value = "carrier", required = false) String carrier,
                                       @RequestParam(value = "funcGroup", required = false) String funcGroup, @RequestParam("station") String station) throws URISyntaxException, IOException {
        log.debug("REST request to save Document : {}", docName);
        Document document = null;
        if (docName != null && !file.isEmpty()) {

            log.debug("Received file {}", file.getOriginalFilename());
            document = new Document(docName, description, showInDashboard, classification, carrier, funcGroup, station);
            documentService.save(document, file);
        }
        return ResponseEntity.created(new URI("/api/documents/upload/" + document.getId())).build();
    }

    /**
     * GET  /documents/download/:file -> get the "file" for document.
     */
    @RequestMapping(value = "/documents/download",
        method = RequestMethod.GET,
        produces = MediaType.ALL_VALUE)
    @Timed
    @Secured(SECTION)
    public void download(@RequestParam("file") Long id, HttpServletRequest request, HttpServletResponse response) throws IOException {
        log.debug("REST request to get Document file id: {}", id);
        // get absolute path of the application
        ServletContext context = request.getServletContext();
        String appPath = context.getRealPath("");
        log.debug("Application file path {}", appPath);
        // construct the complete absolute path of the file
        Document document = documentService.getOne(id);
        String fullPath = appPath + document.getFilePath();
        File downloadFile = new File(fullPath);
        FileInputStream inputStream = new FileInputStream(downloadFile);

        // get MIME type of the file
        String mimeType = document.getFileExtn(); // can get from actual file as well using context
        if (mimeType == null) {
            // set to binary type if MIME mapping not found
            mimeType = "application/octet-stream";
        }
        log.debug("MIME type: {}", mimeType);

        // set content attributes for the response
        response.setContentType(mimeType);
        response.setContentLength((int) downloadFile.length());

        // set headers for the response
        String headerKey = "Content-Disposition";
        String headerValue = String.format("attachment; filename=\"%s\"", downloadFile.getName());
        response.setHeader(headerKey, headerValue);

        // get output stream of the response
        OutputStream outStream = response.getOutputStream();

        byte[] buffer = new byte[BUFFER_SIZE];
        int bytesRead = -1;

        // write bytes read from the input stream into the output stream
        while ((bytesRead = inputStream.read(buffer)) != -1) {
            outStream.write(buffer, 0, bytesRead);
        }

        inputStream.close();
        outStream.close();
    }
```

# Service

```java
@Service
@Transactional
public class DocumentServiceImpl implements DocumentService {

    private final Logger log = LoggerFactory.getLogger(DocumentServiceImpl.class);
    @Inject
    private Environment env;

    @Inject
    private DocumentRepository documentRepository;
    /**
     * System default email address that sends the e-mails.
     */
    private String baseLocation;

    @PostConstruct
    public void init() throws IOException {
        this.baseLocation = env.getProperty("document.path");
        /**
         * Create the base location and all parent paths.
         */
        this.createBaseLocation();
    }



    @Override
    public Document save(Document document, MultipartFile file) throws IOException {

        String filePath = getFolderPath(document);
        String fullFileName = filePath + File.separator + String.valueOf(new Date().getTime()) + "-" + file.getOriginalFilename();
        log.debug("Saving file to disk {}" ,fullFileName);
        createDirectory(getFullFileDirectoryPath(filePath));
        boolean fileSaved = saveFileToDisk(fullFileName, file);
        if(fileSaved){
            document.setFilePath(fullFileName);
            document.setFileSize(file.getSize());
            document.setFileExtn(file.getContentType());
            documentRepository.save(document);
        }
        return document;
    }

    @Override
    public void delete(Long id) throws IOException {
        Document document = documentRepository.getOne(id);
        boolean fileDeleted = deleteFile(document);
        if(fileDeleted){
            documentRepository.delete(id);
        }

    }

    @Override
    @Secured("S_DOCUMENT")
    @SecureRules(method = Method.GET, rules = {Rule.BY_C, Rule.BY_S, Rule.BY_F})
    public Document getOne(Long id) {
        return documentRepository.getOne(id);
    }

    /**
     *
     * @param document
     * @return
     * @throws IOException
     */
    private boolean deleteFile(Document document) throws IOException {
        String fileFullName = getFullFileDirectoryPath(document.getFilePath());
        File file = new File(fileFullName);
        if (file.exists()) {
            try {
                return file.delete();
            } catch (SecurityException e) {
                log.error(e.getMessage(), e);
                throw new IOException("File deletion failed", e);
            }
        }
        return false;
    }

    /**
     *
     * @param document
     * @return
     */
    private String getFolderPath(Document document) {
        return document.getCarrier() + File.separator + document.getStnCode() + File.separator + document.getFuncGroup();
    }

    /**
     *
     * @throws IOException
     */
    private synchronized void createBaseLocation() throws IOException {
        createDirectory(this.baseLocation);

    }

    /**
     * Create the named directory and any parent directories.
     *
     * @param directoryPath
     *            the directory path to create
     *
     * @throws IOException
     */
    private synchronized void createDirectory(String directoryPath)
        throws IOException {

        log.debug("Creating directory {}", directoryPath);
        File dir = new File(directoryPath);

        if (!dir.exists()) {
            try {
                dir.mkdirs();
                return;
            } catch (SecurityException e) {
                log.error(e.getMessage(), e);
                throw new IOException(e);
            }
        } else {
            if (!dir.isDirectory()) {
                /**
                 * It exists but is a file?
                 */
                throw new IOException(String.format(
                    "File name exists and is not a directory: %s",
                    directoryPath));
            }
        }
    }
    /**
     * Write the new chunk to disk and update the Upload record. Don't
     * synchronize here because the file write may take some time. Only
     * synchronize on the update of the Upload.
     *
     * @param path
     * @param file
     *
     * @return
     *
     * @throws IOException
     */
    private boolean saveFileToDisk(String path, MultipartFile file) throws IOException {

        String fullPath = getFullFileDirectoryPath(path);

        File f = getFile(fullPath);

        FileOutputStream output = new FileOutputStream(f, true);

        try {
            output.write(file.getBytes());
        } catch(Exception e){
            log.error("Error saving file to disk {}", e.getMessage(), e);
            throw new IOException("Error saving file to disk", e);
        }finally {
            output.close();
        }

        return true;
    }

    /**
     * Get a File for this filename.
     *
     * @param path
     *
     * @return
     */
    private File getFile(String path) {
        return new File(path);
    }


    /**
     * Given the Upload, get the path that the chunks should be placed in.
     *
     * @param path
     *
     * @return
     */
    private String getFullFileDirectoryPath(String path) {
        return baseLocation + File.separator + path;
    }
}
```

**Comment by deepu105** [comment]
@andidev also on windows this one works when i tested with files upto 1GB , on the spring boot tomcat container. i havent tested above that size
I had to add below to the yml property files to make files more than 20MB work

```
multipart:
    max-file-size: 500Mb
    max-request-size: 1000Mb
```

**Comment by deepu105** [comment]
@andidev the initial snippet i put up had some issues, i have updated with correct one

**Comment by andidev** [comment]
@deepu105 
Thanx, I'll have a look into it when I have some time.

**Comment by deepu105** [comment]
@andidev ng-file-upload has jumped to 7.x we are using 5.x there are some breaking changes and better validation support etc, you may want to check it out

**Comment by andidev** [comment]
Thanks I'll have a look at it as soon as I have some time!

**Comment by andidev** [comment]
I'm not sure but I think we can remove our min and max validation directives and use the ones provided by ngf-file-upload since files are now returned to the ng-model even if an error occured.
This also means that we can easally implement the new validators min and max width/height (at least on front end)

**Comment by coolduebtn** [comment]
is this feature already in the latest release of jhipster or do we need to wait ?

**Comment by andidev** [comment]
It's there since long time ago. Maybe it's called byte. Should be the last field   type .

**Comment by cbornet** [comment]
@coolduebtn Would you be using Cassandra ? Because it is only available for SQL and mongo.

**Comment by ProDoser** [comment]
Hello!
When I used [BETA] Blob data type I've got colmn with type datatype blob in MySQL DB which is  just 65,535 bytes of data. I rsolve this issue by manully change type from `blob` to `mediumblob` in the generated .xml for DB shema.

Would be nice if when Blob-->Image is choosed it will generate `mediumblob` datatype in liquibase .xml

I am totaly new with jhipster, but this is really great tool. Thank you for your effort!

**Comment by andidev** [comment]
@ProDoser 
Are you sure I'm not aware of this issue and most certainly this was tested when it was developed.

**Comment by ProDoser** [comment]
@andidev 
Yes I am sure :) I just modified it. During entity creation I choosed `blob` then `an image` and it was not possible to upload images greater then 64kB. Dadtatype for the column in my DB (MySQL) was` blob` and in the liquibase .xml file it was `<column name="picture" type="blob" >` so I just changed type to `mediumblob` and update my table. Now it works. Not critical, easy to solve, but anyway I just want to point this one. Maybe someone else fall in truble.

**Comment by andidev** [comment]
Did you change to mediumblob in liquibase config or in database? 
I'll fix this ASAP. Though I did not found documentation of the supported datatypes in liquibase. Do you know where I can find them?

**Comment by ProDoser** [comment]
@andidev I'am just learning all this stuff by doing and don't know anythig about liquebase and datatypes yet.

So first time when I tryed to upload an image I've got an exeption. In the logs found that it is not possible to insert data into row because it is too large. Then I checked database and found that datatype for this column is `blob` and according to [W3Schools](http://www.w3schools.com/sql/sql_datatypes.asp) for MySQL `BLOB` up to 65,535 bytes. I tried just to chang datatype in DB and regenerate this entity but did something wrong, probably becaouse removed the .xml and create new one.

Then I basically create new project :) and before first RUN chane in liquibas .xml file for this entity `type="blob"` to `type="mediumblob"` and RUN the application. It created table with `mediumblob` datatype for my image column and now it is possible to upload images more then 64kb.

Maybe in PostgreSQL they have another data types and it works fine, but with MySQL I faced this issue.

**Comment by andidev** [comment]
Yes you are right! There is a issue over here. 
Though it seems like the docs are quite poor for the liquibase blob types. 
Looks like this java source file is the only documentation 
https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/datatype/core/BlobType.java

I think that maybe we need to generate database specific files for each database to be 100% sure that the bkob maxbytes option can be stored in DB. 

I'm creating a new issue for this

**Comment by daddy32** [comment]
Hello.

Sorry for bringing this up again, but I have one issue with this otherwise excellent blob support in JHipster. 

As per design, the REST response is returning all the file contents directly. This is problematic in listing templates - when displaying a list of several records, page is not rendered until the angular client finishes downloading ALL the files from the server - even the query from backend to the database takes quite some time. Even with pagination of the list (depending on the file size), this can take very long. Moreover, the user may not be interested in the file contents at all - he may be looking at other fields of the entity.

Would it be possible to implement some kind of lazy loading for the blob contents? This may be desirable behavior for images which are displayed on the frontend, but may be unnecessary for general blob type.

Thanks for any response.

**Comment by gmarziou** [comment]
@daddy32 please open a new issue for your feature request rather than posting on a PR merged 9 months ago.

**Comment by paulcwarren** [comment]
Hi Folks,

This is an exploratory comment.  

We wanted to draw your attention to [this project](bit.ly/spring-content).  It offers the same level of abstraction for file stores as Spring Data has for databases.  Effectively it is to unstructured data, what Spring Data is to structured data.

In addition to "blobs" and Mongo's GridFS, Spring Content also supports S3 and the good old filesystem.  

Regardless of where it is stored, content can also be fulltext indexed with Apache Solr or Elastic Search and therefore applications can easily offer searching "inside" documents.  It can also render content into different formats making it easy to offer thumbnails for images, extract text from documents and slidedecks or convert office documents into PDFs and vice versa.  

The REST layer also natively support video streaming.

All the features that are required for a serious content management service/application.

At a high level I believe we could make the generators generate code to the generic Spring Content "store" API and control the "where" content is stored through maven/gradle dependencies; i.e. the exact same patterns as your Spring Data generators.   

We wondered whether you would be interested in investigating a joint effort to add Spring Content support to the existing "blob" features.  If so please reply and we can figure out a way to connect.

**Comment by deepu105** [comment]
@paulcwarren sounds interesting but since this js an old PR People might not pay attention. I suggest you open a new issue as feature request and we can discuss there

**Pull request merged**